In [1]:
import pandas as pd
import glob
import shutil
import sys
import numpy as np

In [2]:
# Manual cleaning
    # Removed "Loans that do not meet credit policy"
    # Removed all the lines beginning with "Total amount"
    # All the empty cells made 0
    # Remove first line from all the CSVs. There is some comment.

## Load data
Use read_csv function to load the data. I am already using the dtype for the right kind of assignments. While it takes longer to code it, it helped me later.

In [3]:
# Load all the csv files.
loan_data = pd.DataFrame()
for file in glob.glob("Loan*.csv"):
    print(file)
    df = pd.read_csv(file, dtype={'id':float,  # 0
                                  'member_id':float,  # 1
                                  'loan_amnt':int,  # 2
                                  'funded_amnt':int,  # 3
                                  'funded_amnt_inv':str,  # 4
                                  'term':str,  # 5
                                  'int_rate':str,  # 6
                                  'installment':float,  # 7
                                  'grade':str,  # 8
                                  'sub_grade':str,  # 9
                                  'emp_title':str,  # 10
                                  'emp_length':str,  # 11
                                  'home_ownership':str,  # 12
                                  'annual_inc':str,  # 13
                                  'verification_status':str,  # 14
                                  'issue_d':str,  # 15
                                  'loan_status':str,  # 16
                                  'pymnt_plan':str,  # 17
                                  'url':str,  # 18
                                  'desc':str,  # 19
                                  'purpose':str,  # 20
                                  'title':str,  # 21
                                  'zip_code':str,  # 22
                                  'addr_state':str,  # 23
                                  'dti':float,  # 24
                                  'delinq_2yrs':int,  # 25
                                  'earliest_cr_line':str,  # 26
                                  'inq_last_6mths':int,  # 27
                                  'mths_since_last_delinq':float,  # 28
                                  'mths_since_last_record':float,  # 29
                                  'open_acc':int,  # 30
                                  'pub_rec':int,  # 31
                                  'revol_bal':int,  # 32
                                  'revol_util':str,  # 33
                                  'total_acc':int,  # 34
                                  'initial_list_status':str,  # 35
                                  'out_prncp':float,  # 36
                                  'out_prncp_inv':float,  # 37
                                  'total_pymnt':float,  # 38
                                  'total_pymnt_inv':float,  # 39
                                  'total_rec_prncp':float,  # 40
                                  'total_rec_int':float,  # 41
                                  'total_rec_late_fee':float,  # 42
                                  'recoveries':float,  # 43
                                  'collection_recovery_fee':float,  # 44
                                  'last_pymnt_d':str,  # 45
                                  'last_pymnt_amnt':float,  # 46
                                  'next_pymnt_d':str,  # 47
                                  'last_credit_pull_d':str,  # 48
                                  'collections_12_mths_ex_med':int,  # 49
                                  'mths_since_last_major_derog':float,  # 50
                                  'policy_code':int,  # 51
                                  'application_type':str,  # 52
                                  'annual_inc_joint':float,  # 53
                                  'dti_joint':float,  # 54
                                  'verification_status_joint':str,  # 55
                                  'acc_now_delinq':int,  # 56
                                  'tot_coll_amt':float,  # 57
                                  'tot_cur_bal':float,  # 58
                                  'open_acc_6m':float,  # 59
                                  'open_act_il':float,  # 60
                                  'open_il_12m':float,  # 61
                                  'open_il_24m':float,  # 62
                                  'mths_since_rcnt_il':float,  # 63
                                  'total_bal_il':float,  # 64
                                  'il_util':float,  # 65
                                  'open_rv_12m':float,  # 66
                                  'open_rv_24m':float,  # 67
                                  'max_bal_bc':float,  # 68
                                  'all_util':float,  # 69
                                  'total_rev_hi_lim':float,  # 70
                                  'inq_fi':float,  # 71
                                  'total_cu_tl':float,  # 72
                                  'inq_last_12m':float,  # 73
                                  'acc_open_past_24mths':float,  # 74
                                  'avg_cur_bal':float,  # 75
                                  'bc_open_to_buy':float,  # 76
                                  'bc_util':float,  # 77
                                  'chargeoff_within_12_mths':int,  # 78
                                  'delinq_amnt':int,  # 79
                                  'mo_sin_old_il_acct':float,  # 80
                                  'mo_sin_old_rev_tl_op':float,  # 81
                                  'mo_sin_rcnt_rev_tl_op':float,  # 82
                                  'mo_sin_rcnt_tl':float,  # 83
                                  'mort_acc':float,  # 84
                                  'mths_since_recent_bc':float,  # 85
                                  'mths_since_recent_bc_dlq':float,  # 86
                                  'mths_since_recent_inq':float,  # 87
                                  'mths_since_recent_revol_delinq':float,  # 88
                                  'num_accts_ever_120_pd':float,  # 89
                                  'num_actv_bc_tl':float,  # 90
                                  'num_actv_rev_tl':float,  # 91
                                  'num_bc_sats':float,  # 92
                                  'num_bc_tl':float,  # 93
                                  'num_il_tl':float,  # 94
                                  'num_op_rev_tl':float,  # 95
                                  'num_rev_accts':float,  # 96
                                  'num_rev_tl_bal_gt_0':float,  # 97
                                  'num_sats':float,  # 98
                                  'num_tl_120dpd_2m':float,  # 99
                                  'num_tl_30dpd':float,  # 100
                                  'num_tl_90g_dpd_24m':float,  # 101
                                  'num_tl_op_past_12m':float,  # 102
                                  'pct_tl_nvr_dlq':float,  # 103
                                  'percent_bc_gt_75':float,  # 104
                                  'pub_rec_bankruptcies':float,  # 105
                                  'tax_liens':int,  # 106
                                  'tot_hi_cred_lim':float,  # 107
                                  'total_bal_ex_mort':float,  # 108
                                  'total_bc_limit':float,  # 109
                                  'total_il_high_credit_limit':float,  # 110
                                  'revol_bal_joint':float,  # 111
                                  'sec_app_earliest_cr_line':str,  # 112
                                  'sec_app_inq_last_6mths':float,  # 113
                                  'sec_app_mort_acc':float,  # 114
                                  'sec_app_open_acc':float,  # 115
                                  'sec_app_revol_util':float,  # 116
                                  'sec_app_open_act_il':float,  # 117
                                  'sec_app_num_rev_accts':float,  # 118
                                  'sec_app_chargeoff_within_12_mths':float,  # 119
                                  'sec_app_collections_12_mths_ex_med':float,  # 120
                                  'sec_app_mths_since_last_major_derog':float,  # 121
                                  'hardship_flag':str,  # 122
                                  'hardship_type':str,  # 123
                                  'hardship_reason':str,  # 124
                                  'hardship_status':str,  # 125
                                  'deferral_term':float,  # 126
                                  'hardship_amount':float,  # 127
                                  'hardship_start_date':str,  # 128
                                  'hardship_end_date':str,  # 129
                                  'payment_plan_start_date':str,  # 130
                                  'hardship_length':float,  # 131
                                  'hardship_dpd':float,  # 132
                                  'hardship_loan_status':str,  # 133
                                  'orig_projected_additional_accrued_interest':float,  # 134
                                  'hardship_payoff_balance_amount':float,  # 135
                                  'hardship_last_payment_amount':float,  # 136
                                  'disbursement_method':str,  # 137
                                  'debt_settlement_flag':str,  # 138
                                  'debt_settlement_flag_date':str,  # 139
                                  'settlement_status':str,  # 140
                                  'settlement_date':str,  # 141
                                  'settlement_amount':str,  # 142
                                  'settlement_percentage':str,  # 143
                                  'settlement_term':str})
    # df = pd.read_csv(file, dtype={'loan_amnt':int, 'funded_amnt':int, 'funded_amnt_inv':str, 'term':str, 'int_rate':str, 'installment':float, 'grade':str, 'sub_grade':str, 'emp_title':str, 'emp_length':str, 'home_ownership':str, 'annual_inc':str, 'verification_status':str, 'issue_d':str, 'loan_status':str, 'pymnt_plan':str, 'url':str, 'desc':str, 'purpose':str, 'title':str, 'zip_code':str, 'addr_state':str, 'dti':float, 'delinq_2yrs':int, 'earliest_cr_line':str, 'inq_last_6mths':int, 'mths_since_last_delinq':int, 'mths_since_last_record':int, 'open_acc':int, 'pub_rec':int, 'revol_bal':int, 'revol_util':str, 'total_acc':int, 'initial_list_status':str, 'out_prncp':float, 'out_prncp_inv':float, 'total_pymnt':float, 'total_pymnt_inv':float, 'total_rec_prncp':float, 'total_rec_int':float, 'total_rec_late_fee':int, 'recoveries':int, 'collection_recovery_fee':int, 'last_pymnt_d':str, 'last_pymnt_amnt':float, 'next_pymnt_d':str, 'last_credit_pull_d':str, 'collections_12_mths_ex_med':int, 'mths_since_last_major_derog':int, 'policy_code':int, 'application_type':str, 'annual_inc_joint':int, 'dti_joint':int, 'verification_status_joint':int, 'acc_now_delinq':int, 'tot_coll_amt':int, 'tot_cur_bal':int, 'open_acc_6m':int, 'open_act_il':int, 'open_il_12m':int, 'open_il_24m':int, 'mths_since_rcnt_il':int, 'total_bal_il':int, 'il_util':int, 'open_rv_12m':int, 'open_rv_24m':int, 'max_bal_bc':int, 'all_util':int, 'total_rev_hi_lim':int, 'inq_fi':int, 'total_cu_tl':int, 'inq_last_12m':int, 'acc_open_past_24mths':int, 'avg_cur_bal':int, 'bc_open_to_buy':int, 'bc_util':float, 'chargeoff_within_12_mths':int, 'delinq_amnt':int, 'mo_sin_old_il_acct':int, 'mo_sin_old_rev_tl_op':int, 'mo_sin_rcnt_rev_tl_op':int, 'mo_sin_rcnt_tl':int, 'mort_acc':int, 'mths_since_recent_bc':int, 'mths_since_recent_bc_dlq':int, 'mths_since_recent_inq':int, 'mths_since_recent_revol_delinq':int, 'num_accts_ever_120_pd':int, 'num_actv_bc_tl':int, 'num_actv_rev_tl':int, 'num_bc_sats':int, 'num_bc_tl':int, 'num_il_tl':int, 'num_op_rev_tl':int, 'num_rev_accts':int, 'num_rev_tl_bal_gt_0':int, 'num_sats':int, 'num_tl_120dpd_2m':int, 'num_tl_30dpd':int, 'num_tl_90g_dpd_24m':int, 'num_tl_op_past_12m':int, 'pct_tl_nvr_dlq':float, 'percent_bc_gt_75':float, 'pub_rec_bankruptcies':int, 'tax_liens':int, 'tot_hi_cred_lim':int, 'total_bal_ex_mort':int, 'total_bc_limit':int, 'total_il_high_credit_limit':int, 'revol_bal_joint':int, 'sec_app_earliest_cr_line':int, 'sec_app_inq_last_6mths':int, 'sec_app_mort_acc':int, 'sec_app_open_acc':int, 'sec_app_revol_util':int, 'sec_app_open_act_il':int, 'sec_app_num_rev_accts':int, 'sec_app_chargeoff_within_12_mths':int, 'sec_app_collections_12_mths_ex_med':int, 'sec_app_mths_since_last_major_derog':int, 'hardship_flag':str, 'hardship_type':str, 'hardship_reason':str, 'hardship_status':str, 'deferral_term':int, 'hardship_amount':float, 'hardship_start_date':str, 'hardship_end_date':str, 'payment_plan_start_date':str, 'hardship_length':int, 'hardship_dpd':int, 'hardship_loan_status':str, 'orig_projected_additional_accrued_interest':float, 'hardship_payoff_balance_amount':float, 'hardship_last_payment_amount':float, 'disbursement_method':str, 'debt_settlement_flag':str, 'debt_settlement_flag_date':str, 'settlement_status':str, 'settlement_date':str, 'settlement_amount':str, 'settlement_percentage':str, 'settlement_term':str})
    loan_data = loan_data.append(df)

LoanStats3a.csv
LoanStats3b.csv
LoanStats3c.csv
LoanStats3d.csv
LoanStats_2016Q1.csv
LoanStats_2016Q2.csv
LoanStats_2016Q3.csv
LoanStats_2016Q4.csv
LoanStats_2017Q1.csv
LoanStats_2017Q2.csv
LoanStats_2017Q3.csv
LoanStats_2017Q4.csv


### Basic operations
Get the size of the dataframe. Also, print it. It gives a quick high level view of how the data looks like.

In [4]:
# Delete columns, where all the entries are 0.
loan_data = loan_data.loc[:, (loan_data != 0).any(axis=0)]
print(loan_data.shape)
loan_data.info()

(1765428, 145)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 1765428 entries, 0 to 118647
Columns: 145 entries, id to settlement_term
dtypes: float64(87), int64(14), object(44)
memory usage: 1.9+ GB


In [5]:
print(loan_data)

        id  member_id  loan_amnt  funded_amnt funded_amnt_inv        term  \
0      NaN        0.0       5000         5000            4975   36 months   
1      NaN        0.0       2500         2500            2500   60 months   
2      NaN        0.0       2400         2400            2400   36 months   
3      NaN        0.0      10000        10000           10000   36 months   
4      NaN        0.0       3000         3000            3000   60 months   
5      NaN        0.0       5000         5000            5000   36 months   
6      NaN        0.0       7000         7000            7000   60 months   
7      NaN        0.0       3000         3000            3000   36 months   
8      NaN        0.0       5600         5600            5600   60 months   
9      NaN        0.0       5375         5375            5350   60 months   
10     NaN        0.0       6500         6500            6500   60 months   
11     NaN        0.0      12000        12000           12000   36 months   

### Now read the dictionay also
It has all the information about the various headers of the load_data

In [6]:
# Load the dictionary from Lending Club data
# loan_dict = pd.read_excel(open('LCDataDictionary.xlsx','rb'), sheetname='LoanStats')
loan_dict = pd.read_excel(open('LCDataDictionary.xlsx','rb'), sheet_name='LoanStats')

In [7]:
# Changing max-column-width option for better print.
pd.set_option('max_colwidth',1000)

### Clean data
From this point onwards, read all the data in small chunks and clean it.
Cleaning involves the following steps:
1. Something

In [8]:
# Now analyze one column at a time
# loan_dict.loc[loan_dict['LoanStatNew'] == 'loan_amnt']
print(loan_data.iloc[:5,:6])
for y in loan_data.columns[:6]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[:6]))]

   id  member_id  loan_amnt  funded_amnt funded_amnt_inv        term
0 NaN        0.0       5000         5000            4975   36 months
1 NaN        0.0       2500         2500            2500   60 months
2 NaN        0.0       2400         2400            2400   36 months
3 NaN        0.0      10000        10000           10000   36 months
4 NaN        0.0       3000         3000            3000   60 months
id float64
member_id float64
loan_amnt int64
funded_amnt int64
funded_amnt_inv object
term object


LoanStatNew  \
23      funded_amnt   
24  funded_amnt_inv   
27               id   
41        loan_amnt   
44        member_id   
94             term   

                                                                                                                                                                     Description  
23                                                                                                                The total amount committed to that loan at that point in time.  
24                                                                                                  The total amount committed by investors for that loan at that point in time.  
27                                                                                                                                 A unique LC assigned ID for the loan listing.  
41  The listed amount of the loan applied for by the borrower. If at some point in time, the credit department reduces the loan amount, then it will be reflected in this value.  
44                                                                                                                              A unique LC assigned Id for the borrower member.  
94                                                                                          The number of payments on the loan. Values are in months and can be either 36 or 60.

In [9]:
# Analysis
# loan_amnt is int64 - Good
# funded_amnt is int64 - Good
# funded_amnt_inv is object - Bad
# term is object - Bad
# int_rate is object - Bad

# Let's start with funded_amnt_inv
# Find the non-int entry
df = loan_data['funded_amnt_inv'].to_frame().applymap(np.isreal) == False
df.loc[df['funded_amnt_inv'] == True]

funded_amnt_inv
0                  True
1                  True
2                  True
3                  True
4                  True
5                  True
6                  True
7                  True
8                  True
9                  True
10                 True
11                 True
12                 True
13                 True
14                 True
15                 True
16                 True
17                 True
18                 True
19                 True
20                 True
21                 True
22                 True
23                 True
24                 True
25                 True
26                 True
27                 True
28                 True
29                 True
...                 ...
118618             True
118619             True
118620             True
118621             True
118622             True
118623             True
118624             True
118625             True
118626             True
118627             True
118628             True
118629             True
118630             True
118631             True
118632             True
118633             True
118634             True
118635             True
118636             True
118637             True
118638             True
118639             True
118640             True
118641             True
118642             True
118643             True
118644             True
118645             True
118646             True
118647             True

[1765428 rows x 1 columns]

In [10]:
# Looks like all the entries are considered non-integral
# Let's examine
loan_data['funded_amnt_inv'][:1]
# Since each object seems to string, convert it to numeric.
loan_data['funded_amnt_inv'] = pd.to_numeric(loan_data['funded_amnt_inv'])
# Find the non-numeric entry
df = loan_data['funded_amnt_inv'].to_frame().applymap(np.isreal) == False
df.loc[df['funded_amnt_inv'] == True]
df = loan_data['funded_amnt_inv']
df.loc[df.isnull()]
# So now the data for funded_amnt_inv is all good.

Series([], Name: funded_amnt_inv, dtype: float64)

In [11]:
# Moving to "term"
# All terms are described in terms of months.
# So, a simple replacement should work.
loan_data['term'] = loan_data['term'].str.replace(" months","")
# Now convert all of them to numeric
loan_data['term'] = pd.to_numeric(loan_data['term'])
# Find the non-numeric entry
df = loan_data['term'].to_frame().applymap(np.isreal) == False
df.loc[df['term'] == True]
# The term data looks all good.

Empty DataFrame
Columns: [term]
Index: []

In [12]:
# Moving to "int_rate"
# All int_rate are described in terms of %.
# So, a simple replacement should work.
loan_data['int_rate'] = loan_data['int_rate'].str.replace("%","")
# Now convert all of them to numeric
loan_data['int_rate'] = pd.to_numeric(loan_data['int_rate'])
# Find the non-numeric entry
df = loan_data['int_rate'].to_frame().applymap(np.isreal) == False
df.loc[df['int_rate'] == True]
# The int_rate data looks all good.

Empty DataFrame
Columns: [int_rate]
Index: []

In [13]:
# Moving to "installment"
# Find the non-numeric entry
df = loan_data['installment'].to_frame().applymap(np.isreal) == False
df.loc[df['installment'] == True]
# The installment data looks all good.

Empty DataFrame
Columns: [installment]
Index: []

In [14]:
# Print thoese features again
print(loan_data.iloc[:5,:6])
for y in loan_data.columns[:6]:
  print(y, loan_data[y].dtype)

   id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  term
0 NaN        0.0       5000         5000           4975.0    36
1 NaN        0.0       2500         2500           2500.0    60
2 NaN        0.0       2400         2400           2400.0    36
3 NaN        0.0      10000        10000          10000.0    36
4 NaN        0.0       3000         3000           3000.0    60
id float64
member_id float64
loan_amnt int64
funded_amnt int64
funded_amnt_inv float64
term int64


In [15]:
# Want to see if the loan_amnt and funded_amnt are ever different
df_ = loan_data['loan_amnt'] - loan_data['funded_amnt'] != 0
df = df_.to_frame()
print(len(df.loc[df[0] == True]))

2065


Since, there are 2065 cases where the loan_amnt and funded_amnt are different I am going to keep both the features.
2065 is really small compared to the 1.7 million data we have.
But, for now, I will keep it.

In [16]:
# Look for 0 loan_amnt
loan_data.drop(loan_data[loan_data.loan_amnt == 0].index, inplace=True)
loan_data.loc[loan_data['loan_amnt'] == 0]

Empty DataFrame
Columns: [id, member_id, loan_amnt, funded_amnt, funded_amnt_inv, term, int_rate, installment, grade, sub_grade, emp_title, emp_length, home_ownership, annual_inc, verification_status, issue_d, loan_status, pymnt_plan, url, desc, purpose, title, zip_code, addr_state, dti, delinq_2yrs, earliest_cr_line, inq_last_6mths, mths_since_last_delinq, mths_since_last_record, open_acc, pub_rec, revol_bal, revol_util, total_acc, initial_list_status, out_prncp, out_prncp_inv, total_pymnt, total_pymnt_inv, total_rec_prncp, total_rec_int, total_rec_late_fee, recoveries, collection_recovery_fee, last_pymnt_d, last_pymnt_amnt, next_pymnt_d, last_credit_pull_d, collections_12_mths_ex_med, mths_since_last_major_derog, policy_code, application_type, annual_inc_joint, dti_joint, verification_status_joint, acc_now_delinq, tot_coll_amt, tot_cur_bal, open_acc_6m, open_act_il, open_il_12m, open_il_24m, mths_since_rcnt_il, total_bal_il, il_util, open_rv_12m, open_rv_24m, max_bal_bc, all_util, total_rev_hi_lim, inq_fi, total_cu_tl, inq_last_12m, acc_open_past_24mths, avg_cur_bal, bc_open_to_buy, bc_util, chargeoff_within_12_mths, delinq_amnt, mo_sin_old_il_acct, mo_sin_old_rev_tl_op, mo_sin_rcnt_rev_tl_op, mo_sin_rcnt_tl, mort_acc, mths_since_recent_bc, mths_since_recent_bc_dlq, mths_since_recent_inq, mths_since_recent_revol_delinq, num_accts_ever_120_pd, num_actv_bc_tl, num_actv_rev_tl, num_bc_sats, num_bc_tl, num_il_tl, num_op_rev_tl, num_rev_accts, num_rev_tl_bal_gt_0, num_sats, num_tl_120dpd_2m, ...]
Index: []

[0 rows x 145 columns]

In [17]:
# Now analyze the next 5 features
print(loan_data.iloc[:5,6:11])
for y in loan_data.columns[6:11]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[6:11]))]

   int_rate  installment grade sub_grade                 emp_title
0     10.65       162.87     B        B2                         0
1     15.27        59.83     C        C4                     Ryder
2     15.96        84.33     C        C5                         0
3     13.49       339.31     C        C1       AIR RESOURCES BOARD
4     12.69        67.79     B        B5  University Medical Group
int_rate float64
installment float64
grade object
sub_grade object
emp_title object


LoanStatNew  \
20    emp_title   
25        grade   
33  installment   
34     int_rate   
92    sub_grade   

                                                            Description  
20  The job title supplied by the Borrower when applying for the loan.*  
25                                               LC assigned loan grade  
33     The monthly payment owed by the borrower if the loan originates.  
34                                            Interest Rate on the loan  
92                                            LC assigned loan subgrade

In [18]:
# Starting with grade
loan_data['grade'].unique()

array(['B', 'C', 'A', 'E', 'F', 'D', 'G'], dtype=object)

In [19]:
# 'grade' looks good for a categorical type.
# The 0 corresponds to no data.
# For now, it is ok in out case.

# Next try sub_grade
loan_data['sub_grade'].unique()

array(['B2', 'C4', 'C5', 'C1', 'B5', 'A4', 'E1', 'F2', 'C3', 'B1', 'D1',
       'A1', 'B3', 'B4', 'C2', 'D2', 'A3', 'A5', 'D5', 'A2', 'E4', 'D3',
       'D4', 'F3', 'E3', 'F4', 'F1', 'E5', 'G4', 'E2', 'G3', 'G2', 'G1',
       'F5', 'G5'], dtype=object)

In [20]:
# Again 'sub_grade' looks decent

# Moving to emp_title
print(len(loan_data['emp_title'].unique()))
loan_data['emp_title'].unique()

449772


array(['0', 'Ryder', 'AIR RESOURCES BOARD', ..., 'Acft mechanic',
       'Content Review ', 'BABYSITTER'], dtype=object)

There 449772 employment title including '0', which means no data. Hence, I am removing it from the loan data table.

In [21]:
# Too many of them.
# Will drop 'emp_title' from the loan_data for now.
# Will later do some analysis if needed.
loan_data = loan_data.drop('emp_title', 1)

In [22]:
# Moving to emp_length
# First remove "year", "years" and blanks
loan_data['emp_length'] = loan_data['emp_length'].str.replace("years","")
loan_data['emp_length'] = loan_data['emp_length'].str.replace("year","")
loan_data['emp_length'] = loan_data['emp_length'].str.replace(" ","")
print(loan_data['emp_length'].unique())
# Replacing all NaN with 0.
# Not sure, if this is the right way.
loan_data.loc[(loan_data['emp_length'] != '<1') & \
              (loan_data['emp_length'] != '1') & \
              (loan_data['emp_length'] != '2') & \
              (loan_data['emp_length'] != '3') & \
              (loan_data['emp_length'] != '4') & \
              (loan_data['emp_length'] != '5') & \
              (loan_data['emp_length'] != '6') & \
              (loan_data['emp_length'] != '7') & \
              (loan_data['emp_length'] != '8') & \
              (loan_data['emp_length'] != '9') & \
              (loan_data['emp_length'] != '10+'), 'emp_length'] = 0
print(loan_data['emp_length'].unique())

['10+' '<1' '1' '3' '8' '9' '4' '5' '6' '2' '7' nan]
['10+' '<1' '1' '3' '8' '9' '4' '5' '6' '2' '7' 0]


In [23]:
# Moving to home_ownership
loan_data['home_ownership'].unique()

array(['RENT', 'OWN', 'MORTGAGE', 'OTHER', 'NONE', 'ANY'], dtype=object)

In [24]:
# Though OTHER, NONE and ANY are a little confusing, I will do that cleaning later.
# For now things look ok.

# Printing all the columngs together again.
print(loan_data.iloc[:5,6:10])

   int_rate  installment grade sub_grade
0     10.65       162.87     B        B2
1     15.27        59.83     C        C4
2     15.96        84.33     C        C5
3     13.49       339.31     C        C1
4     12.69        67.79     B        B5


In [25]:
# Moving to next 6 columns
print(loan_data.iloc[:5,10:15])
for y in loan_data.columns[10:15]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[10:15]))]

  emp_length home_ownership annual_inc verification_status issue_d
0        10+           RENT      24000            Verified  Dec-11
1         <1           RENT      30000     Source Verified  Dec-11
2        10+           RENT      12252        Not Verified  Dec-11
3        10+           RENT      49200     Source Verified  Dec-11
4          1           RENT      80000     Source Verified  Dec-11
emp_length object
home_ownership object
annual_inc object
verification_status object
issue_d object


LoanStatNew  \
4             annual_inc   
19            emp_length   
26        home_ownership   
35               issue_d   
112  verification_status   

                                                                                                                                               Description  
4                                                                            The self-reported annual income provided by the borrower during registration.  
19                      Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years.   
26   The home ownership status provided by the borrower during registration or obtained from the credit report. Our values are: RENT, OWN, MORTGAGE, OTHER  
35                                                                                                                     The month which the loan was funded  
112                                                             Indicates if income was verified by LC, not verified, or if the income source was verified

In [26]:
# Staring with annual_inc
# It should be numeric
# So, let'f find the non-numeric entry
df = loan_data['annual_inc'].to_frame().applymap(np.isreal) == False
len(df.loc[df['annual_inc'] == True])

1765404

In [27]:
# Looks like all the entries are considered non-integral
# Let's examine
loan_data['annual_inc'][:1]
# Since each object seems to string, convert it to numeric.
loan_data['annual_inc'] = pd.to_numeric(loan_data['annual_inc'])
# Find the non-numeric entry
df = loan_data['annual_inc'].to_frame().applymap(np.isreal) == False
df.loc[df['annual_inc'] == True]
df = loan_data['annual_inc']
df.loc[df.isnull()]
# So now the data for annual_inc is all good.

Series([], Name: annual_inc, dtype: float64)

In [28]:
# Now verification_status
loan_data['verification_status'].unique()

array(['Verified', 'Source Verified', 'Not Verified'], dtype=object)

In [29]:
# issue_d
loan_data['issue_d'].unique()

array(['Dec-11', 'Nov-11', 'Oct-11', 'Sep-11', 'Aug-11', 'Jul-11',
       'Jun-11', 'May-11', 'Apr-11', 'Mar-11', 'Feb-11', 'Jan-11',
       'Dec-10', 'Nov-10', 'Oct-10', 'Sep-10', 'Aug-10', 'Jul-10',
       'Jun-10', 'May-10', 'Apr-10', 'Mar-10', 'Feb-10', 'Jan-10',
       'Dec-09', 'Nov-09', 'Oct-09', 'Sep-09', 'Aug-09', 'Jul-09',
       'Jun-09', 'May-09', 'Apr-09', 'Mar-09', 'Feb-09', 'Jan-09',
       'Dec-08', 'Nov-08', 'Oct-08', 'Sep-08', 'Aug-08', 'Jul-08',
       'Jun-08', 'May-08', 'Apr-08', 'Mar-08', 'Feb-08', 'Jan-08',
       'Dec-07', 'Nov-07', 'Oct-07', 'Sep-07', 'Aug-07', 'Jul-07',
       'Jun-07', 'Dec-2013', 'Nov-2013', 'Oct-2013', 'Sep-2013',
       'Aug-2013', 'Jul-2013', 'Jun-2013', 'May-2013', 'Apr-2013',
       'Mar-2013', 'Feb-2013', 'Jan-2013', 'Dec-2012', 'Nov-2012',
       'Oct-2012', 'Sep-2012', 'Aug-2012', 'Jul-2012', 'Jun-2012',
       'May-2012', 'Apr-2012', 'Mar-2012', 'Feb-2012', 'Jan-2012',
       'Dec-2014', 'Nov-2014', 'Oct-2014', 'Sep-2014', 'Aug-2014

Breakdown the date into two. One will have the year and the other will have the month.

In [30]:
def try_convert_date(value):
    try:
        new_val = pd.datetime.strptime(value, '%b-%y')
    except:
        new_val = pd.datetime.strptime(value, '%b-%Y')
    return new_val

# df[0] = pd.to_datetime(df[0], format='%b-%y')

loan_data['issue_d_month'] = pd.DatetimeIndex(loan_data['issue_d'].apply(try_convert_date)).month
loan_data['issue_d_year'] = pd.DatetimeIndex(loan_data['issue_d'].apply(try_convert_date)).year
print(loan_data['issue_d_month'].unique())
print(loan_data['issue_d_year'].unique())

[12 11 10  9  8  7  6  5  4  3  2  1]
[2011 2010 2009 2008 2007 2013 2012 2014 2015 2016 2017]


In [31]:
# Drop issue_d
loan_data = loan_data.drop('issue_d', 1)

In [32]:
# loan_status
loan_data['loan_status'].unique()

array(['Fully Paid', 'Charged Off',
       'Does not meet the credit policy. Status:Fully Paid',
       'Does not meet the credit policy. Status:Charged Off', 'Current',
       'Late (16-30 days)', 'Late (31-120 days)', 'In Grace Period',
       'Default', 'Issued'], dtype=object)

In [33]:
# pymnt_plan
loan_data['pymnt_plan'].unique()

array(['n', 'y'], dtype=object)

In [34]:
# Moving to next 6 columns
# Dropping description out of hte list as it is pure text.
# Will see later, if adding it adds any value.
loan_data = loan_data.drop('desc', 1)
print(loan_data.iloc[:5,15:20])
for y in loan_data.columns[15:20]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[15:20]))]

  pymnt_plan url         purpose                 title zip_code
0          n   0     credit_card              Computer    860xx
1          n   0             car                  bike    309xx
2          n   0  small_business  real estate business    606xx
3          n   0           other              personel    917xx
4          n   0           other              Personal    972xx
pymnt_plan object
url object
purpose object
title object
zip_code object


LoanStatNew  \
87      purpose   
88   pymnt_plan   
95        title   
111         url   
114    zip_code   

                                                                               Description  
87                              A category provided by the borrower for the loan request.   
88                          Indicates if a payment plan has been put in place for the loan  
95                                                 The loan title provided by the borrower  
111                                                 URL for the LC page with listing data.  
114  The first 3 numbers of the zip code provided by the borrower in the loan application.

In [35]:
# url is the URL for the LC page with listing data.
# This data is not needed for our purpose
loan_data = loan_data.drop('url', 1)

In [36]:
# Let't check purpose
loan_data['purpose'].unique()

array(['credit_card', 'car', 'small_business', 'other', 'wedding',
       'debt_consolidation', 'home_improvement', 'major_purchase',
       'medical', 'moving', 'vacation', 'house', 'renewable_energy',
       'educational'], dtype=object)

In [37]:
# 'purpose' looks good.
# Let's try title
len(loan_data['title'].unique())

63150

In [38]:
# Again, 'title' is text entry which, for now, I am going to remove
loan_data = loan_data.drop('title', 1)

In [39]:
# zip_code
len(loan_data['zip_code'].unique())

954

In [40]:
# I am going to keep zip_code.
# The number of unique values are not too bad.
# Have to see how many zip_codes where missing and I replaced with 0.
len(loan_data.loc[loan_data['zip_code'] == 0])
# Since only one entry has the zip_code 0, I am going to randomly assign that item with one of the existing zip_codes.
# I hope, this won't affect things bad.
import random
loan_data.loc[(loan_data['zip_code'] == 0), 'zip_code'] =  random.choice(loan_data['zip_code'].unique())
len(loan_data.loc[loan_data['zip_code'] == 0])

0

In [41]:
# Moving to addr_state
loan_data['addr_state'].unique()

array(['AZ', 'GA', 'IL', 'CA', 'OR', 'NC', 'TX', 'VA', 'MO', 'CT', 'UT',
       'FL', 'NY', 'PA', 'MN', 'NJ', 'KY', 'OH', 'SC', 'RI', 'LA', 'MA',
       'WA', 'WI', 'AL', 'CO', 'KS', 'NV', 'AK', 'MD', 'WV', 'VT', 'MI',
       'DC', 'SD', 'NH', 'AR', 'NM', 'MT', 'HI', 'WY', 'OK', 'DE', 'MS',
       'TN', 'IA', 'NE', 'ID', 'IN', 'ME', 'ND'], dtype=object)

In [42]:
# Moving to next 6 columns
print(loan_data.iloc[:6,18:23])
for y in loan_data.columns[18:23]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[18:23]))]

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths
0         AZ  27.65            0           Jan-85               1
1         GA   1.00            0           Apr-99               5
2         IL   8.72            0           Nov-01               2
3         CA  20.00            0           Feb-96               1
4         OR  17.94            0           Jan-96               0
5         AZ  11.20            0           Nov-04               3
addr_state object
dti float64
delinq_2yrs int64
earliest_cr_line object
inq_last_6mths int64


LoanStatNew  \
2         addr_state   
13       delinq_2yrs   
16               dti   
18  earliest_cr_line   
32    inq_last_6mths   

                                                                                                                                                                                                 Description  
2                                                                                                                                                 The state provided by the borrower in the loan application  
13                                                                                              The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years  
16  A ratio calculated using the borrower’s total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower’s self-reported monthly income.  
18                                                                                                                                         The month the borrower's earliest reported credit line was opened  
32                                                                                                                          The number of inquiries in past 6 months (excluding auto and mortgage inquiries)

In [43]:
# dti is float, delinq_2yrs is int and inq_last_6mths is int.
# They look good.
# mts_since_last_delinq is float.
# Let's check it.
# Ideally, it should have been int
loan_data['mths_since_last_delinq'].unique()

array([  0.,  35.,  38.,  61.,   8.,  20.,  18.,  68.,  45.,  48.,  41.,
        40.,  74.,  25.,  53.,  39.,  10.,  26.,  56.,  77.,  28.,  52.,
        24.,  16.,  60.,  54.,  23.,   9.,  11.,  13.,  65.,  19.,  80.,
        22.,  59.,  79.,  44.,  64.,  57.,  14.,  63.,  49.,  15.,  73.,
        70.,  29.,  51.,   5.,  75.,  55.,   2.,  30.,  47.,  33.,  69.,
         4.,  43.,  21.,  27.,  46.,  81.,  78.,  82.,  31.,  76.,  62.,
        72.,  42.,  50.,   3.,  12.,  67.,  36.,  34.,  58.,  17.,  71.,
        66.,  32.,   6.,  37.,   7.,   1.,  83.,  86., 115.,  96., 103.,
       120., 106.,  89., 107.,  85.,  97.,  95.,  nan, 110.,  84., 135.,
        88.,  87., 122.,  91., 146., 134., 114.,  99.,  93., 127., 101.,
        94., 102., 129., 113., 139., 131., 156., 143., 109., 119., 149.,
       118., 130., 148., 126.,  90., 141., 116., 100., 152.,  98.,  92.,
       108., 133., 104., 111., 105., 170., 124., 136., 180., 188., 140.,
       151., 159., 121., 123., 157., 112., 154., 17

In [62]:
# The data looks clean.
# I can safely convert it to int.

# Convert all 0s and nan to 0.
# loan_data.loc[loan_data['mths_since_last_delinq'].isnull()]['mths_since_last_delinq'] = 0
loan_data[['mths_since_last_delinq']] = loan_data[['mths_since_last_delinq']].applymap(lambda x: 0 if pd.isnull(x) else x)
print(loan_data['mths_since_last_delinq'].unique())

loan_data[['mths_since_last_delinq']] = loan_data[['mths_since_last_delinq']].astype(int)
loan_data['mths_since_last_delinq'].unique()

[  0.  35.  38.  61.   8.  20.  18.  68.  45.  48.  41.  40.  74.  25.
  53.  39.  10.  26.  56.  77.  28.  52.  24.  16.  60.  54.  23.   9.
  11.  13.  65.  19.  80.  22.  59.  79.  44.  64.  57.  14.  63.  49.
  15.  73.  70.  29.  51.   5.  75.  55.   2.  30.  47.  33.  69.   4.
  43.  21.  27.  46.  81.  78.  82.  31.  76.  62.  72.  42.  50.   3.
  12.  67.  36.  34.  58.  17.  71.  66.  32.   6.  37.   7.   1.  83.
  86. 115.  96. 103. 120. 106.  89. 107.  85.  97.  95. 110.  84. 135.
  88.  87. 122.  91. 146. 134. 114.  99.  93. 127. 101.  94. 102. 129.
 113. 139. 131. 156. 143. 109. 119. 149. 118. 130. 148. 126.  90. 141.
 116. 100. 152.  98.  92. 108. 133. 104. 111. 105. 170. 124. 136. 180.
 188. 140. 151. 159. 121. 123. 157. 112. 154. 171. 142. 125. 117. 176.
 137. 145. 158. 192. 132. 168. 195. 161. 128. 178. 160. 150. 202. 153.
 138. 162.]


array([  0,  35,  38,  61,   8,  20,  18,  68,  45,  48,  41,  40,  74,
        25,  53,  39,  10,  26,  56,  77,  28,  52,  24,  16,  60,  54,
        23,   9,  11,  13,  65,  19,  80,  22,  59,  79,  44,  64,  57,
        14,  63,  49,  15,  73,  70,  29,  51,   5,  75,  55,   2,  30,
        47,  33,  69,   4,  43,  21,  27,  46,  81,  78,  82,  31,  76,
        62,  72,  42,  50,   3,  12,  67,  36,  34,  58,  17,  71,  66,
        32,   6,  37,   7,   1,  83,  86, 115,  96, 103, 120, 106,  89,
       107,  85,  97,  95, 110,  84, 135,  88,  87, 122,  91, 146, 134,
       114,  99,  93, 127, 101,  94, 102, 129, 113, 139, 131, 156, 143,
       109, 119, 149, 118, 130, 148, 126,  90, 141, 116, 100, 152,  98,
        92, 108, 133, 104, 111, 105, 170, 124, 136, 180, 188, 140, 151,
       159, 121, 123, 157, 112, 154, 171, 142, 125, 117, 176, 137, 145,
       158, 192, 132, 168, 195, 161, 128, 178, 160, 150, 202, 153, 138,
       162])

In [ ]:
# All looks good

# earliest_cr_line is type object, which is good.
print('[%s]' % ', '.join(map(str,  loan_data['earliest_cr_line'].unique() )))

In [ ]:
# Moving to next 6 columns
print(loan_data.iloc[:6,23:28])
for y in loan_data.columns[23:28]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[23:28]))]

In [ ]:
# Start analyzing mths_since_last_record
loan_data['mths_since_last_record'].unique()

In [ ]:
# Look good and all of them are integers (as there are no partial months).
# Hence, converting all of them to int
loan_data[['mths_since_last_record']] = loan_data[['mths_since_last_record']].astype(int)
loan_data['mths_since_last_record'].unique()

In [ ]:
# Start analyzing open_acc
loan_data['open_acc'].unique()

In [ ]:
# Start analyzing pub_rec
loan_data['pub_rec'].unique()

In [ ]:
# Start analyzing revol_bal
loan_data['revol_bal'].unique()

In [ ]:
# Start analyzing revol_util
# All revol_util are described in terms of %.
# So, a simple replacement should work.
loan_data['revol_util'] = loan_data['revol_util'].str.replace("%","")
# Now convert all of them to numeric
loan_data['revol_util'] = pd.to_numeric(loan_data['revol_util'])
# Find the non-numeric entry
df = loan_data['revol_util'].to_frame().applymap(np.isreal) == False
df.loc[df['revol_util'] == True]
# The revol_util data looks all good.

In [ ]:
# Moving to next 5 columns
print(loan_data.iloc[:6,28:33])
for y in loan_data.columns[28:33]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[28:33]))]

In [ ]:
# Start analyzing total_acc
loan_data['total_acc'].unique()

In [ ]:
# Start analyzing initial_list_status
loan_data['initial_list_status'].unique()

In [ ]:
# Start analyzing out_prncp
loan_data['out_prncp'].unique()

In [ ]:
# Start analyzing out_prncp_inv
loan_data['out_prncp_inv'].unique()

In [ ]:
# Want to see if the out_prncp and out_prncp_inv are ever different
df_ = loan_data['out_prncp'] - loan_data['out_prncp_inv'] != 0
df = df_.to_frame()
print(len(df.loc[df[0] == True]))

In [ ]:
# Since, there is a non-zero number of cases where the the out_prncp and out_prncp_inv are different, I am keeping it.
# Though, 45K entries out of 1.7million seems small.o

# Start analyzing total_pymnt
loan_data['total_pymnt'].unique()

In [ ]:
# Moving to next 4 columns
print(loan_data.iloc[:6,33:37])
for y in loan_data.columns[33:37]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[33:37]))]

In [ ]:
# Moving to next 4 columns
print(loan_data.iloc[:6,37:41])
for y in loan_data.columns[37:41]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[37:41]))]

In [ ]:
# All except look good in terms of the data type
# Cheching last_pymnt_d
loan_data['last_pymnt_d'].unique()

In [ ]:
import re
# There are inconsistencies.
# Some places the year is called 16 and other places it is 2016
loan_data['last_pymnt_d'] = loan_data['last_pymnt_d'].str.replace("-20","")
loan_data['last_pymnt_d'].replace(to_replace="([a-zA-Z])(\d+)", value=r"\1-\2", regex=True, inplace=True)
loan_data['last_pymnt_d'].unique()

In [ ]:
# Delete the entries with last_pymnt_d = 0, if the total number of those are less than 0.1% of the total data.
# Else, we need to think of a way to fix it.
if (len(loan_data.loc[loan_data['last_pymnt_d'] == '0']) < len(loan_data) * 0.1 / 100):
    loan_data.drop(loan_data[loan_data.last_pymnt_d == '0'].index, inplace=True)
    loan_data.loc[loan_data['last_pymnt_d'] == '0']
loan_data['last_pymnt_d'].unique()

In [ ]:
# Moving to next 4 columns
print(loan_data.iloc[:6,41:44])
for y in loan_data.columns[41:44]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[41:44]))]

In [ ]:
# next_pymnt_d looks wrong
len(loan_data.loc[loan_data['next_pymnt_d'] == '0'])
# It has 92648 entries.
# For now, I am dropping this. 
# Next payment is future payment is not completely relevant.
# This information should already by hidden in all the other data.
loan_data = loan_data.drop('next_pymnt_d', 1)

In [ ]:
# Analyze last_credit_pull_d
loan_data['last_credit_pull_d'].unique()

In [ ]:
# There are inconsistencies.
# Some places the year is called 16 and other places it is 2016
loan_data['last_credit_pull_d'] = loan_data['last_credit_pull_d'].str.replace("-20","")
loan_data['last_credit_pull_d'].replace(to_replace="([a-zA-Z])(\d+)", value=r"\1-\2", regex=True, inplace=True)
loan_data['last_credit_pull_d'].unique()

In [ ]:
# I am keeping the 0 values in the last_credit_pull_d.
# It could mean, the credit was not pulled before.

loan_data['collections_12_mths_ex_med'].unique()

In [ ]:
# collections_12_mths_ex_med looks clean.
# Moving on

In [ ]:
# Moving to next 4 columns
print(loan_data.iloc[:6,44:48])
for y in loan_data.columns[44:48]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[44:48]))]

In [ ]:
# mths_since_last_major_derog should be int
loan_data[['mths_since_last_major_derog']] = loan_data[['mths_since_last_major_derog']].astype(int)
loan_data['mths_since_last_major_derog'].unique()

In [ ]:
loan_data['policy_code'].unique()

In [ ]:
# Value for policy_code is all 1.
# Dropping it.
loan_data = loan_data.drop('policy_code', 1)

In [ ]:
loan_data['application_type'].unique()

In [ ]:
loan_data['annual_inc_joint'].unique()

In [ ]:
# Joint annual income of 0 looks like missing data.
len(loan_data.loc[loan_data['annual_inc_joint'] == 0.0])

In [ ]:
# THat is large number of data, which is showing annual_inc_joint = 0.
# And there is no easy way to fix it.
# So, dropping it.
loan_data = loan_data.drop('annual_inc_joint', 1)

In [ ]:
# Moving to next 4 columns
print(loan_data.iloc[:6,47:51])
for y in loan_data.columns[47:51]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[47:51]))]

In [ ]:
loan_data['dti_joint'].unique()

In [ ]:
len(loan_data.loc[loan_data['dti_joint'] == 0.0])

In [ ]:
# The dti_joint = 0 is a large percentage of our total data.
# Hence, dropping it.
loan_data = loan_data.drop('dti_joint', 1)

In [ ]:
loan_data['verification_status_joint'].unique()

In [ ]:
len(loan_data.loc[loan_data['verification_status_joint'] == '0']) + len(loan_data.loc[loan_data['verification_status_joint'] == 0.0])

In [ ]:
# Again a big chunk of the verification_status_joint is 0.
# Hence, dropping it.
loan_data = loan_data.drop('verification_status_joint', 1)

In [ ]:
loan_data['acc_now_delinq'].unique()

In [ ]:
loan_data['tot_coll_amt'].unique()

In [ ]:
# Moving to next 4 columns
print(loan_data.iloc[:6,49:54])
for y in loan_data.columns[49:54]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[49:54]))]

In [ ]:
# The datatypes of all of them look good. May need to change the type to int for tot_cur_bal
loan_data['tot_cur_bal'].unique()

In [ ]:
# Converting tot_cur_bal to int as it won't do any harm
loan_data[['tot_cur_bal']] = loan_data[['tot_cur_bal']].astype(int)

In [ ]:
# Coverting all the others also to int
loan_data[['open_acc_6m']] = loan_data[['open_acc_6m']].astype(int)
loan_data[['open_act_il']] = loan_data[['open_act_il']].astype(int)
loan_data[['open_il_12m']] = loan_data[['open_il_12m']].astype(int)
loan_data[['open_il_24m']] = loan_data[['open_il_24m']].astype(int)

In [ ]:
# Now print to see if there is anything I missed.
print(loan_data['open_acc_6m'].unique())
print(loan_data['open_il_12m'].unique())
print(loan_data['open_il_24m'].unique())
print(loan_data['open_act_il'].unique())
print(loan_data['tot_cur_bal'].unique())

In [ ]:
# Moving to next 4 columns
print(loan_data.iloc[:6,59:64])
for y in loan_data.columns[59:64]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[59:64]))]

In [ ]:
# all of these should be int type
loan_data[['max_bal_bc']] = loan_data[['max_bal_bc']].astype(int)
loan_data[['all_util']] = loan_data[['all_util']].astype(int)
loan_data[['total_rev_hi_lim']] = loan_data[['total_rev_hi_lim']].astype(int)
loan_data[['inq_fi']] = loan_data[['inq_fi']].astype(int)
loan_data[['total_cu_tl']] = loan_data[['total_cu_tl']].astype(int)

In [ ]:
# And print
print(loan_data['max_bal_bc'].unique())
print(loan_data['all_util'].unique())
print(loan_data['total_rev_hi_lim'].unique())
print(loan_data['inq_fi'].unique())
print(loan_data['total_cu_tl'].unique())

In [ ]:
# Moving to next 4 columns
print(loan_data.iloc[:6,64:69])
for y in loan_data.columns[64:69]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[64:69]))]

In [ ]:
# Other than bc_util, everything can be int.
loan_data[['acc_open_past_24mths']] = loan_data[['acc_open_past_24mths']].astype(int)
loan_data[['avg_cur_bal']] = loan_data[['avg_cur_bal']].astype(int)
loan_data[['bc_open_to_buy']] = loan_data[['bc_open_to_buy']].astype(int)
loan_data[['inq_last_12m']] = loan_data[['inq_last_12m']].astype(int)

print(loan_data['acc_open_past_24mths'].unique())
print(loan_data['avg_cur_bal'].unique())
print(loan_data['bc_open_to_buy'].unique())
print(loan_data['inq_last_12m'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,69:72])
for y in loan_data.columns[69:72]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[69:72]))]

In [ ]:
# Converting mo_sin_old_il_acct to int
loan_data[['inq_last_12m']] = loan_data[['inq_last_12m']].astype(int)
loan_data['inq_last_12m'].unique()

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,72:76])
for y in loan_data.columns[72:76]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[72:76]))]

In [ ]:
# Converting all of them to int
loan_data[['mo_sin_rcnt_tl']] = loan_data[['mo_sin_rcnt_tl']].astype(int)
loan_data[['mort_acc']] = loan_data[['mort_acc']].astype(int)
loan_data[['mths_since_recent_bc']] = loan_data[['mths_since_recent_bc']].astype(int)
loan_data[['mths_since_recent_bc_dlq']] = loan_data[['mths_since_recent_bc_dlq']].astype(int)

print(loan_data['mo_sin_rcnt_tl'].unique())
print(loan_data['mort_acc'].unique())
print(loan_data['mths_since_recent_bc'].unique())
print(loan_data['mths_since_recent_bc_dlq'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,76:78])
for y in loan_data.columns[76:78]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[76:78]))]

In [ ]:
# Converting all of them to int
loan_data[['mths_since_recent_inq']] = loan_data[['mths_since_recent_inq']].astype(int)
loan_data[['mths_since_recent_revol_delinq']] = loan_data[['mths_since_recent_revol_delinq']].astype(int)

print(loan_data['mths_since_recent_inq'].unique())
print(loan_data['mths_since_recent_revol_delinq'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,78:82])
for y in loan_data.columns[78:82]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[78:82]))]

In [ ]:
# Converting all of them to int
loan_data[['num_accts_ever_120_pd']] = loan_data[['num_accts_ever_120_pd']].astype(int)
loan_data[['num_actv_bc_tl']] = loan_data[['num_actv_bc_tl']].astype(int)
loan_data[['num_actv_rev_tl']] = loan_data[['num_actv_rev_tl']].astype(int)
loan_data[['num_bc_sats']] = loan_data[['num_bc_sats']].astype(int)

print(loan_data['num_accts_ever_120_pd'].unique())
print(loan_data['num_actv_bc_tl'].unique())
print(loan_data['num_actv_rev_tl'].unique())
print(loan_data['num_bc_sats'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,82:87])
for y in loan_data.columns[82:87]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[82:87]))]

In [ ]:
# Converting all of them to int
loan_data[['num_bc_tl']] = loan_data[['num_bc_tl']].astype(int)
loan_data[['num_il_tl']] = loan_data[['num_il_tl']].astype(int)
loan_data[['num_op_rev_tl']] = loan_data[['num_op_rev_tl']].astype(int)
loan_data[['num_rev_accts']] = loan_data[['num_rev_accts']].astype(int)
loan_data[['num_rev_tl_bal_gt_0']] = loan_data[['num_rev_tl_bal_gt_0']].astype(int)

print(loan_data['num_bc_tl'].unique())
print(loan_data['num_il_tl'].unique())
print(loan_data['num_op_rev_tl'].unique())
print(loan_data['num_rev_accts'].unique())
print(loan_data['num_rev_tl_bal_gt_0'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,87:91])
for y in loan_data.columns[87:91]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[87:91]))]

In [ ]:
# Converting all of them to int
loan_data[['num_sats']] = loan_data[['num_sats']].astype(int)
loan_data[['num_tl_120dpd_2m']] = loan_data[['num_tl_120dpd_2m']].astype(int)
loan_data[['num_tl_30dpd']] = loan_data[['num_tl_30dpd']].astype(int)
loan_data[['num_tl_90g_dpd_24m']] = loan_data[['num_tl_90g_dpd_24m']].astype(int)

print(loan_data['num_sats'].unique())
print(loan_data['num_tl_120dpd_2m'].unique())
print(loan_data['num_tl_30dpd'].unique())
print(loan_data['num_tl_90g_dpd_24m'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,91:95])
for y in loan_data.columns[91:95]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[91:95]))]

In [ ]:
# converting some to int
loan_data[['num_tl_op_past_12m']] = loan_data[['num_tl_op_past_12m']].astype(int)
loan_data[['pub_rec_bankruptcies']] = loan_data[['pub_rec_bankruptcies']].astype(int)

print(loan_data['num_tl_op_past_12m'].unique())
print(loan_data['pct_tl_nvr_dlq'].unique())
print(loan_data['percent_bc_gt_75'].unique())
print(loan_data['pub_rec_bankruptcies'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,95:99])
for y in loan_data.columns[95:99]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[95:99]))]

In [ ]:
# Converting all of them to in
loan_data[['tot_hi_cred_lim']] = loan_data[['tot_hi_cred_lim']].astype(int)
loan_data[['total_bal_ex_mort']] = loan_data[['total_bal_ex_mort']].astype(int)
loan_data[['total_bc_limit']] = loan_data[['total_bc_limit']].astype(int)

print(loan_data['tax_liens'].unique())
print(loan_data['tot_hi_cred_lim'].unique())
print(loan_data['total_bal_ex_mort'].unique())
print(loan_data['total_bc_limit'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,99:102])
for y in loan_data.columns[99:102]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[99:102]))]

In [ ]:
print(loan_data['total_il_high_credit_limit'].unique())
print(loan_data['revol_bal_joint'].unique())
print(loan_data['sec_app_earliest_cr_line'].unique())

In [ ]:
#Converting the numeric data to in
loan_data[['total_il_high_credit_limit']] = loan_data[['total_il_high_credit_limit']].astype(int)
loan_data[['revol_bal_joint']] = loan_data[['revol_bal_joint']].astype(int)

In [ ]:
# Show all the sec_app_earliest_cr_line, where the data is 0
print(len(loan_data.loc[loan_data['sec_app_earliest_cr_line'] == '0']))
print(len(loan_data.loc[loan_data['sec_app_earliest_cr_line'] == 0]))
print(len(loan_data.loc[loan_data['sec_app_earliest_cr_line'] != 0]))

# Converting all the 0s to one type
loan_data.loc[loan_data['sec_app_earliest_cr_line'] == 0, 'sec_app_earliest_cr_line'] = '0'

In [ ]:
# Fow now, I am keeping all of them.
# Later, I will do histogram and see if there is somethign else, I should do.

# Moving to next columns
print(loan_data.iloc[:6,102:105])
for y in loan_data.columns[102:105]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[102:105]))]

In [ ]:
print(loan_data['sec_app_inq_last_6mths'].unique())
print(loan_data['sec_app_mort_acc'].unique())
print(loan_data['sec_app_open_acc'].unique())

In [ ]:
# These can be safely converted to int
loan_data[['sec_app_inq_last_6mths']] = loan_data[['sec_app_inq_last_6mths']].astype(int)
loan_data[['sec_app_mort_acc']] = loan_data[['sec_app_mort_acc']].astype(int)
loan_data[['sec_app_open_acc']] = loan_data[['sec_app_open_acc']].astype(int)

print(loan_data['sec_app_inq_last_6mths'].unique())
print(loan_data['sec_app_mort_acc'].unique())
print(loan_data['sec_app_open_acc'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,105:108])
for y in loan_data.columns[105:108]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[105:108]))]

In [ ]:
# First print the values. If things look ok, then convert them to int.
print(loan_data['sec_app_revol_util'].unique())
print(loan_data['sec_app_open_act_il'].unique())
print(loan_data['sec_app_num_rev_accts'].unique())

In [ ]:
# Convert them to int
loan_data[['sec_app_revol_util']] = loan_data[['sec_app_revol_util']].astype(int)
loan_data[['sec_app_open_act_il']] = loan_data[['sec_app_open_act_il']].astype(int)
loan_data[['sec_app_num_rev_accts']] = loan_data[['sec_app_num_rev_accts']].astype(int)

print(loan_data['sec_app_revol_util'].unique())
print(loan_data['sec_app_open_act_il'].unique())
print(loan_data['sec_app_num_rev_accts'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,108:110])
for y in loan_data.columns[108:110]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[108:110]))]

In [ ]:
# Print
print('Before')
print(loan_data['sec_app_chargeoff_within_12_mths'].unique())
print(loan_data['sec_app_collections_12_mths_ex_med'].unique())

# Convert them to int
loan_data[['sec_app_chargeoff_within_12_mths']] = loan_data[['sec_app_chargeoff_within_12_mths']].astype(int)
loan_data[['sec_app_collections_12_mths_ex_med']] = loan_data[['sec_app_collections_12_mths_ex_med']].astype(int)

# Print again
print('After')
print(loan_data['sec_app_chargeoff_within_12_mths'].unique())
print(loan_data['sec_app_collections_12_mths_ex_med'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,110:113])
for y in loan_data.columns[110:113]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[110:113]))]

In [ ]:
print(loan_data['sec_app_mths_since_last_major_derog'].unique())
print(loan_data['hardship_flag'].unique())
print(loan_data['hardship_type'].unique())

In [ ]:
# Converting sec_app_mths_since_last_major_derog to int.
loan_data[['sec_app_mths_since_last_major_derog']] = loan_data[['sec_app_mths_since_last_major_derog']].astype(int)

# Dropping hardship_type as there is only one type of non-zero data
loan_data = loan_data.drop('hardship_type', 1)

print(loan_data['sec_app_mths_since_last_major_derog'].unique())
print(loan_data['hardship_flag'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,112:116])
for y in loan_data.columns[112:116]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[112:116]))]

In [ ]:
print(loan_data['hardship_reason'].unique())
print(loan_data['hardship_status'].unique())
print(loan_data['deferral_term'].unique())
print(loan_data['hardship_amount'].unique())

In [ ]:
# Converting some of them to int
loan_data[['deferral_term']] = loan_data[['deferral_term']].astype(int)

# Converting all the 0s to one type
loan_data.loc[loan_data['hardship_reason'] == 0, 'hardship_reason'] = '0'
loan_data.loc[loan_data['hardship_status'] == 0, 'hardship_status'] = '0'

print(loan_data['hardship_reason'].unique())
print(loan_data['hardship_status'].unique())
print(loan_data['deferral_term'].unique())
print(loan_data['hardship_amount'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,116:119])
for y in loan_data.columns[116:119]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[116:119]))]

In [ ]:
# Converting all the 0s to one type
loan_data.loc[loan_data['hardship_start_date'] == 0, 'hardship_start_date'] = '0'
loan_data.loc[loan_data['hardship_end_date'] == 0, 'hardship_end_date'] = '0'
loan_data.loc[loan_data['payment_plan_start_date'] == 0, 'payment_plan_start_date'] = '0'

print(loan_data['hardship_start_date'].unique())
print(loan_data['hardship_end_date'].unique())
print(loan_data['payment_plan_start_date'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,119:122])
for y in loan_data.columns[119:122]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[119:122]))]

In [ ]:
print(loan_data['hardship_length'].unique())
print(loan_data['hardship_dpd'].unique())
print(loan_data['hardship_loan_status'].unique())

In [ ]:
# Converting some to int
loan_data['hardship_length'] = loan_data['hardship_length'].astype(int)
loan_data['hardship_dpd'] = loan_data['hardship_dpd'].astype(int)

print(loan_data['hardship_length'].unique())
print(loan_data['hardship_dpd'].unique())
print(loan_data['hardship_loan_status'].unique())

In [ ]:
# Converting all the 0s to one type
loan_data.loc[loan_data['hardship_loan_status'] == 0, 'hardship_loan_status'] = '0'
print(loan_data['hardship_length'].unique())
print(loan_data['hardship_dpd'].unique())
print(loan_data['hardship_loan_status'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,122:124])
for y in loan_data.columns[122:124]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[122:124]))]

In [ ]:
print(loan_data['orig_projected_additional_accrued_interest'].unique())
print(loan_data['hardship_payoff_balance_amount'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,124:127])
for y in loan_data.columns[124:127]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[124:127]))]

In [ ]:
print(loan_data['hardship_last_payment_amount'].unique())
print(loan_data['disbursement_method'].unique())
print(loan_data['debt_settlement_flag'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,127:130])
for y in loan_data.columns[127:130]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[127:130]))]

In [ ]:
print(loan_data['debt_settlement_flag_date'].unique())
print(loan_data['settlement_status'].unique())
print(loan_data['settlement_date'].unique())

In [ ]:
# Fixing date format
loan_data['debt_settlement_flag_date'] = loan_data['debt_settlement_flag_date'].str.replace("-20","")
loan_data['debt_settlement_flag_date'].replace(to_replace="([a-zA-Z])(\d+)", value=r"\1-\2", regex=True, inplace=True)
print(loan_data['debt_settlement_flag_date'].unique())

loan_data['settlement_date'] = loan_data['settlement_date'].str.replace("-20","")
loan_data['settlement_date'].replace(to_replace="([a-zA-Z])(\d+)", value=r"\1-\2", regex=True, inplace=True)
print(loan_data['settlement_date'].unique())

# Converting all the 0s to one type
loan_data.loc[loan_data['settlement_status'] == 0, 'settlement_status'] = '0'
print(loan_data['settlement_status'].unique())

In [ ]:
# Moving to next columns
print(loan_data.iloc[:6,130:])
for y in loan_data.columns[130:]:
  print(y, loan_data[y].dtype)
loan_dict.loc[loan_dict['LoanStatNew'].isin(list(loan_data.columns.values[130:]))]

In [ ]:
print(loan_data['settlement_amount'].unique())
print(loan_data['settlement_percentage'].unique())
print(loan_data['settlement_term'].unique())

In [ ]:
# Since each object seems to string, convert it to numeric.
loan_data['settlement_amount'] = pd.to_numeric(loan_data['settlement_amount'])
loan_data['settlement_percentage'] = pd.to_numeric(loan_data['settlement_percentage'])
loan_data['settlement_term'] = pd.to_numeric(loan_data['settlement_term'])

print(loan_data['settlement_amount'].unique())
print(loan_data['settlement_percentage'].unique())
print(loan_data['settlement_term'].unique())

In [ ]:
# At this point the preliminary data reading and conversion are done.
# Save data pickle form
# Next would be to read the pickle data and run histogram in each entry.

loan_data.to_pickle('loan_data.pkl')